`cond` -> `case` where appropriate

alexandergunnarson opened the pull request:
- Moving from `cond` to `case` when possible turns a worst-case O(N)
  (where N is the number of possible objects to which the test constant
  might be equal) check into a worst-case O(1) check.
- While it is not clear whether the affected code paths are
  bottlenecks, performance will nevertheless be improved in those code
  paths.
- All tests continue to pass

alexandergunnarson committed:
`cond` -> `case` where appropriate

- Moving from `cond` to `case` when possible turns a worst-case O(N)
(where N is the number of possible objects to which the test constant
might be equal) check into a worst-case O(1) check.
- While it is not clear whether the affected code paths are
bottlenecks, performance will nevertheless be improved in those code
paths.
- All tests continue to pass

mpenet commented on src/mesomatic/types.clj:
```diff
@@ -99,12 +99,12 @@
-    (= status Protos$Status/DRIVER_ABORTED)     :driver-aborted
-    (= status Protos$Status/DRIVER_STOPPED)     :driver-stopped
-    :else status))
+  (case status
```
I think `case` doesn't work like this for enum values, as they are not compile time literals. 

see:

```
(case java.util.concurrent.TimeUnit/DAYS java.util.concurrent.TimeUnit/DAYS 1 :broken)
-> :broken

```

You could compare ordinal enum values, that's something often macro'ified (https://github.com/mpenet/commons/blob/master/src/clj/qbits/commons/enum.clj#L28-L38). 

In the same file I mentioned I have utils to automate this kind of things (ex create a fn that takes clojure style enum keyword (lowercased "_/-" replaced, etc) and throw with a meaningful error message), feel free to lift these functions if you need them. 

you can see it in use here: https://github.com/mpenet/alia/blob/master/modules/alia/src/qbits/alia/enum.clj

alexandergunnarson committed:
`case` -> dispatch map for enums

Following a comment by @mpenet , `case` dispatch on enums was changed
to a dispatch map whose keys are enums, which is accessed via a call to
`get`. Not O(1) like `case`, but still O(log32(N)), which is quite
comparable.

alexandergunnarson commented:
@mpenet — thanks for your comment! While I myself would have used your `case-enum`, I didn't feel that @pyr would like the extra dependency thrown in, so I changed the instances of `case` which used enums to instead doing something like the following:

```
(def ^:private abc-dispatch-map
  {class.subclass.Enum$A :a
   class.subclass.Enum$B :b
   class.subclass.Enum$C :c
   class.subclass.Enum$D :d})

(defn abc [enum]
  (get abc-dispatch-map enum enum))
```

Not as pretty, but it avoids the dependency.

If @pyr is open to bringing in alia as a dependency, then I'd be more than happy to change those instances to `case-enum`.

mpenet commented:
You re welcome. You can just copy paste de code you need btw, most of  this stuff is derivative work anyway.

alexandergunnarson commented:
All right haha. I'll do that then. Just a sec

alexandergunnarson committed:
dispatch-map -> `case-enum`

As per @mpenet , added/pasted in the `case-enum` function and avoided
the dispatch maps. Looks cleaner.

alexandergunnarson commented:
All right, I committed the `case-enum` edits. I wish there were a way to determine the type of `case` needed at compile-time (e.g. raw `case` or `case-enum`, or maybe something else)... I'm almost certain it's possible. Oh well — those acrobatics aren't needed here.

pyr commented:
Hi @alexandergunnarson, sorry for not acting on this sooner, if you're ok with rebasing against master, i'll merge this and release shortly.

alexandergunnarson committed:
Merge remote-tracking branch 'pyr/master'

# Conflicts:
#	src/mesomatic/types.clj

alexandergunnarson commented:
Hey @pyr — that's all right. Just rebased against master and CI is running. Thanks for looking into this PR!

pyr commented:
Thanks a lot @alexandergunnarson !

alexandergunnarson commented:
No problem at all @pyr !
